server/src/main/java/org/opensearch/gateway/AsyncBatchShardFetch.java

```diff
@@ -0,0 +1,488 @@
+
+    /**
+     * An action that lists the relevant shard data that needs to be fetched.
+     */
+    public interface Lister<NodesResponse extends BaseNodesResponse<NodeResponse>, NodeResponse extends BaseNodeResponse> {
+        void list(DiscoveryNode[] nodes,Map<ShardId, String> shardIdsWithCustomDataPath, ActionListener<NodesResponse> listener);
+    }
+
+    protected final Logger logger;
+    protected final String type;
+    private final String batchUUID;
+    protected Map<ShardId,String> shardsToCustomDataPathMap;
+    private Map<ShardId, Set<String>> ignoredShardToNodes = new HashMap<>();
+    private final AsyncBatchShardFetch.Lister<BaseNodesResponse<T>, T> action;
+    private final Map<String, AsyncBatchShardFetch.NodeEntry<T>> cache = new HashMap<>();
+    private final Set<String> nodesToIgnore = new HashSet<>();
+    private final AtomicLong round = new AtomicLong();
+    private boolean closed;
+
+    @SuppressWarnings("unchecked")
+    protected AsyncBatchShardFetch(
+        Logger logger,
+        String type,
+        Map<ShardId, String> shardsToCustomDataPathMap,
+        AsyncBatchShardFetch.Lister<? extends BaseNodesResponse<T>, T> action,
+        String batchUUID
+    ) {
+        this.logger = logger;
+        this.type = type;
+        this.shardsToCustomDataPathMap = shardsToCustomDataPathMap;
+        this.action = (AsyncBatchShardFetch.Lister<BaseNodesResponse<T>, T>) action;
+        this.batchUUID = batchUUID;
+    }
+
+    @Override
+    public synchronized void close() {
+        this.closed = true;
+    }
+
+    /**
+     * Returns the number of async fetches that are currently ongoing.
+     */
+    public synchronized int getNumberOfInFlightFetches() {
+        int count = 0;
+        for (AsyncBatchShardFetch.NodeEntry<T> nodeEntry : cache.values()) {
+            if (nodeEntry.isFetching()) {
+                count++;
+            }
+        }
+        return count;
+    }
+
+    /**
+     * Fetches the data for the relevant shard. If there any ongoing async fetches going on, or new ones have
+     * been initiated by this call, the result will have no data.
+     * <p>
+     * The ignoreNodes are nodes that are supposed to be ignored for this round, since fetching is async, we need
+     * to keep them around and make sure we add them back when all the responses are fetched and returned.
+     */
+    public synchronized AsyncBatchShardFetch.FetchResult<T> fetchData(DiscoveryNodes nodes, Map<ShardId, Set<String>> ignoreNodes) {
+        if (closed) {
+            throw new IllegalStateException(batchUUID + ": can't fetch data on closed async fetch");
+        }
+
+        // create a flat map for all ignore nodes in a batch.
+        // This information is only used for retrying. Fetching is still a broadcast to all available nodes
+        nodesToIgnore.addAll(ignoreNodes.values().stream().flatMap(Collection::stream).collect(Collectors.toSet()));
```

Comment:
Don't think we need to create a flat map here. Ignore nodes are per shard basis. And we want to keep that data structure same in RoutingAllocation object.

So, if we create a flat map here and pass in result, we'll end up updating the ignore nodes union (of all shards in this batch) for all the shard in this batch. 

But I see you're only using the correct map in `processAllocation` method, so better removing this flat map.

Reply:
nodesToIgnore is used for retrying at line 160, so for a batch we still be needing a flatmap, if a shard in batch contains ignoreNodes we will have safe reroute called batched.